ppb_vector/vector2.py

```diff
@@ -1,97 +1,174 @@
 
 
-_fakevector = collections.namedtuple('_fakevector', ['x', 'y'])
+@functools.lru_cache()
+def _find_lowest_type(left: typing.Type, right: typing.Type) -> typing.Type:
+    """
+    Guess which is the more specific type.
+    """
+    # Basically, see what classes are unique in each type's MRO and return who
+    # has the most.
+    lmro = set(left.__mro__)
+    rmro = set(right.__mro__)
+    lspecial = lmro - rmro
+    rspecial = rmro - lmro
+    if len(lmro) > len(rmro):
+        return left
+    elif len(rmro) > len(lmro):
+        return right
+    else:
+        # They're equal, just arbitrarily pick one
+        return left
+
 
-def _mkvector(value, *, castto=_fakevector):
-    if isinstance(value, Vector2):
-        return value
-    # FIXME: Allow all types of sequences
-    elif isinstance(value, (list, tuple)) and len(value) == 2:
-        return castto(value[0], value[1])
-    # FIXME: Allow all types of mappings
-    elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
-        return castto(value['x'], value['y'])
+def _find_lowest_vector(left: typing.Type, right: typing.Type) -> typing.Type:
+    if left is right:
+        return left
+    elif not issubclass(left, Vector2):
+        return right
+    elif not issubclass(right, Vector2):
+        return left
     else:
-        raise ValueError(f"Cannot use {value} as a vector-like")
+        return _find_lowest_type(left, right)
 
 
-class Vector2(Sequence):
+class Vector2:
+    x: float
+    y: float
+    _length: float
 
-    def __init__(self, x: Real, y: Real):
-        self.x = x
-        self.y = y
-        self.length = hypot(x, y)
+    def __init__(self, x: typing.SupportsFloat, y: typing.SupportsFloat):
+        try:
+            self.x = x.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(x).__name__} object not convertable to float")
+        try:
+            self.y = y.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(y).__name__} object not convertable to float")
 
     @classmethod
-    def convert(cls, value: VectorLike) -> 'Vector2':
+    def convert(cls: typing.Type[VectorOrSub], value: VectorLike) -> VectorOrSub:
         """
         Constructs a vector from a vector-like.
         """
-        return _mkvector(value, castto=type(cls))
+        # Use Vector2.convert() instead of type(self).convert() so that 
+        # _find_lowest_vector() can resolve things well.
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, Vector2):
+            return cls(value.x, value.y)
+        # FIXME: Allow all types of sequences
+        elif isinstance(value, (list, tuple)) and len(value) == 2:
+            return cls(value[0].__float__(), value[1].__float__())
+        # FIXME: Allow all types of mappings
+        elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
```

Comment:
`isinstance(value, typing.Mapping)`?

Also, do we want to enforce `len(value) == 2`?
It might be useful to be able to build a vector from a dict like `{'x': 0, 'y': 1, 'color': 'brown'}`.

Reply:
For now, I think we want to keep exact forms to prevent mistakes.

Especially since afaik no code actually uses the dict form.

Reply:
You added `Mapping` to the imports in a6161b9ea916fd893b4493fcf29d2b747ca1f360 but you forgot to do anything with it.

Reply:
I did, this comment is outdated.